### PR TITLE
Add `cg.SycamoreTargetGateset` and deprecate `cg.ConvertToSycamoreGates`

### PR DESCRIPTION
**Pull request by tanujkhattar**
- Part of https://github.com/quantumlib/Cirq/issues/5028
- Follows the new Transformer API https://github.com/quantumlib/Cirq/issues/4483
- Supports no compile tags NoCompile Tag for optimizers https://github.com/quantumlib/Cirq/issues/4253

### CODE REVIEW

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            if op1.gate == cirq.SWAP and isinstance(op2.gate, cirq.ZZPowGate):
+                return True
+        return False
+
+    tags_to_ignore = context.tags_to_ignore if context else ()
+    circuit = cirq.merge_operations_to_circuit_op(
+        circuit,
+        merge_func_swap_rzz,
+        tags_to_ignore=tags_to_ignore,
+        merged_circuit_op_tag=merged_swap_rzz_tag,
```
Comment (verult) [inline cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py:77]: Why do we tag these connected components? Is it for users to read the circuit and identify transformation steps?

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+
+def test_sycamore_gateset_compiles_swap_zz():
+    qubits = cirq.LineQubit.range(3)
+
+    gamma = np.random.randn()
+    circuit1 = cirq.Circuit(
+        cirq.SWAP(qubits[0], qubits[1]),
+        cirq.Z(qubits[2]),
+        cirq.ZZ(qubits[0], qubits[1]) ** gamma,
```
Comment (verult) [inline cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:107]: Does differing values of `gamma` change the the compiled circuit structure drastically? Wondering why you chose not to assert the compiled circuit against a hard-coded circuit.

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    q1 = cirq.LineQubit(1)
+    subcircuit = cirq.FrozenCircuit(UnknownGate()(q0, q1))
+    circuit = cirq.Circuit(cirq.CircuitOperation(subcircuit))
+    with pytest.raises(ValueError, match='Unable to convert'):
+        cirq.optimize_for_target_gateset(
+            circuit, gateset=cirq_google.SycamoreTargetGateset(), ignore_failures=False
+        )
+
+
+def test_unsupported_gate_ignoring_failures():
```
Comment (verult) [inline cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:192]: nit: Add comment explaining why it's OK to ignore the failure.

**Review by verult — COMMENTED** [review]
lgtm other than some minor comments

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            ops.SYC,
+            cirq.MeasurementGate,
+            cirq.PhasedXZGate,
+            cirq.PhasedXPowGate,
+            cirq.XPowGate,
+            cirq.YPowGate,
+            cirq.ZPowGate,
+            name='SycamoreTargetGateset',
+        )
+        if tabulation is not None and not isinstance(tabulation, cirq.TwoQubitGateTabulation):
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py]: Removed the assertion, it got copied over from the previous implementation that had a similar flow.

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    """Target gateset containing Sycamore + single qubit rotations + Measurement gates."""
+
+    def __init__(
+        self, *, atol: float = 1e-8, tabulation: Optional[cirq.TwoQubitGateTabulation] = None
+    ) -> None:
+        """Inits `cirq_google.SycamoreTargetGateset`.
+
+        Args:
+            atol: A limit on the amount of absolute error introduced by the decomposition.
+            tabulation: If set, a tabulation for the Sycamore gate to use for decomposing Matrix
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py]: Reworded the docstring to reflect this better.

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py (hunk 1 above)
Comment (tanujkhattar) [reply cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py:78]: Yes, the tags are useful in logging but also useful for subsequent transformation steps where we want to identify whether a newly created connected component is a swap + rzz component (and hence is a known decomposition) or is an arbitrary 2q connected component and should be decomposed analytically or using the tabulation.

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:107]: The tests were copied from previous implementation of convert_to_sycamore_gates.py and I don't think changing gamma to a hardcoded value would make much of a difference. 

The important thing to test is that swap + rzz does get compiled via the known compilations flow.

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py (hunk 3 above)
Comment (tanujkhattar) [reply cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:192]: Ignoring failures can be useful in a variety of different use cases. For example, when this method occurs is called as part of a larger transformer pipeline and the user has subsequent transformation steps to convert any unsupported gates. 

I think the flag name + args are clear enough to describe what they do and answering the "why" is beyond the scope here.

**Comment by tanujkhattar** [comment]
@verult Addressed all your comments. Please let me know if you have any further feedback or else approve the PR so I can submit. Thanks for the review!

File: cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py (hunk 3 above)
Comment (verult) [reply cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset_test.py:192]: Ah my confusion was around how "ignoring failures" is being tested, but realized later that `ignore_failures` defaults to True in `optimize_for_target_gateset()` :) Explicitly setting the parameter in the call here would be helpful in reading the test case, but again very much a nit.

**Review by verult — APPROVED** [review]

**Pull request merged**